sklearn/calibration.py

```diff
@@ -236,28 +255,73 @@ def fit(self, X, y, sample_weight=None):
                     warnings.warn("Since %s does not support sample_weights, "
                                   "sample weights will only be used for the "
                                   "calibration itself." % estimator_name)
+            if self.ensemble:
+                # Check that each cross-validation fold can have at least one
+                # example per class
+                if isinstance(self.cv, int):
+                    n_folds = self.cv
+                elif hasattr(self.cv, "n_splits"):
+                    n_folds = self.cv.n_splits
+                else:
+                    n_folds = None
+                if n_folds and np.any([np.sum(y == class_) < n_folds
+                                       for class_ in self.classes_]):
+                    raise ValueError(f"Requesting {n_folds}-fold "
+                                     "cross-validation but provided less than "
+                                     f"{n_folds} examples for at least one "
+                                     "class.")
+                cv = check_cv(self.cv, y, classifier=True)
+
+                for train, test in cv.split(X, y):
+                    this_estimator = clone(base_estimator)
+
+                    if (sample_weight is not None
+                            and base_estimator_supports_sw):
+                        this_estimator.fit(X[train], y[train],
+                                           sample_weight=sample_weight[train])
+                    else:
+                        this_estimator.fit(X[train], y[train])
+
+                    sw = None if sample_weight is None else sample_weight[test]
+                    calibrated_classifier = _fit_calibrator(
+                        this_estimator, self.label_encoder_, self.method,
+                        y=y[test], X=X[test], sample_weight=sw
+                    )
+                    self.calibrated_classifiers_.append(calibrated_classifier)
+            else:
+                if hasattr(base_estimator, "decision_function"):
+                    base_estimator_method = "decision_function"
+                elif hasattr(base_estimator, "predict_proba"):
+                    base_estimator_method = "predict_proba"
+                else:
+                    raise RuntimeError("'base_estimator' as no "
+                                       "'decision_function' or 'predict_proba'"
+                                       " method.")
+                df = cross_val_predict(base_estimator, X, y, cv=cv,
+                                       method=base_estimator_method)
+                if base_estimator_method == "decision_function":
+                    if df.ndim == 1:
+                        df = df[:, np.newaxis]
+                else:
+                    if len(self.label_encoder_.classes_) == 2:
+                        df = df[:, 1:]
```

Comment:
The duplication is a bit unfortunate, there should be a way to merge this logic with that of `_get_predictions`

Maybe by extracting this into a `_reshape_df` utility?
Or maybe we can even let `_get_predictions` call `cross_val_predict` somehow, so that we can also merge the method checks above. 
Also is `df = df[:, 1:]` just equivalent to `df = df[:, 1]`?

Reply:
> Or maybe we can even let `_get_predictions` call `cross_val_predict` somehow, so that we can also merge the method checks above.

This is what I was trying to do before - but `_get_predictions` is is used within `_fit_calibrator` which is within the cv for loop. This is why I wanted to put the cv for loop within `_get_predictions` but it got complicated.

I will add a `_reshape_df` function. Also I have a little dislike for the `df` naming as it is not always 'decision_function`. (I also first think df=dataframe but maybe that is just me)

Reply:
Also I can't think of an elegant way to add the `_reshape_df` function as we get the predictions of the classifier in `_get_predictions` but we use `cross_val_predict` to get the predictions in the case above...

Reply:
Yes I was confused by `df` too. We can rename it as `predictions` or just `preds`.

Reply:
> Also I can't think of an elegant way to add the _reshape_df

Would it help to also add a `_get_method` helper, like

```py
def _get_method(clf):
	# return decision_function (the actual method) or predict_proba or raise error
```

?

Reply:
Well funny you said that because I was just looking into using [`_check_classifier_response_method`](https://github.com/scikit-learn/scikit-learn/blob/6c0e04a7e3d156d1b4246b74b8c0fd66296fbd78/sklearn/metrics/_plot/base.py#L6) (from the plotting module) which does just that.
The thing is that inside `_get_predictions` we get the preds first, before we reshape it. Whereas above, we already have the preds and we just need to reshape. I could use `_check_classifier_response_method` to get the method, get the preds if necessary, then use the `reshape_perds` function to reshape. Is this too complicated?

Reply:
I'm not sure about using ` _check_classifier_response_method ` because it gives higher priority to `predict_proba`.

One way to do it would be something like:

```py
def get_prediction_method(clf, return_string=False):
	# return clf.decision_function or clf.predict_proba or their corresponding string
	# we need strings when using cross_val_predict

def _get_predictions(X, pred_func):
	# pred_func is either a method of the clf, or a partial cross_val_predict
	preds = pred_func(X)
	# reshape preds here
	return preds

# and then we'd  call _get_predictions with one of:
_get_predictions(X, get_prediction_method(clf, return_string=False)  # when ensemble=True
_get_predictions(X, partial(cross_val_predict(..., method=get_prediction_method(clf, return_string=True)))  # when ensemble=False

```

If you think this looks too complex, let's just have a `_reshape_preds` function.

> The thing is that inside _get_predictions we (need to) get the preds first, before we reshape it. Whereas above, we already have the preds and we just need to reshape

I think that's fine. At execution this would indeed repeat the checks `if hasattr...` twice, but that's crazy cheap anyway and it would still make the code clearer IMO (unless there's something I don't see)

Reply:
> I'm not sure about using `_check_classifier_response_method` because it gives higher priority to `predict_proba`

You're right, I thought you could pick priority.

> If you think this looks too complex, let's just have a _reshape_preds function.

I can see how that would work but it is a little complicated (for me, as I'm not familiar with `partial`). Is the advantage 'readability' ?

Reply:
the advantage is no code duplication and succinctness. If it comes at the cost of being too hard to read / maintain, it's not worth it though. You'll be the judge.

partials are a classic functional-programming tool but I guess they're more of an intermediate / advanced concept. We do use them, mostly in the tests though

Reply:
Looking into it, `_fit_calibrator` returns a `_CalibratedClassifierPipeline` instance for each cv fold. For `ensemble=False`, this won't work (for how `_fit_calibrator` works atm) - we want to fit once with all the predictions. We would need to take `_fit_calibrator` out of the cv for loop.

Reply:
> Looking into it, _fit_calibrator returns a _CalibratedClassifierPipeline instance for each cv fold

Only when ensemble=True, right?

There's no CV loop when `ensemble=False`, the "loop" is in `cross_val_predict` and we only need to call `_fit_calibrator` once. Maybe I misunderstood?

Reply:
This is difficult over text! I may be confused too. 

At the moment we have:
```Python
class CalibratedClassifierCV():

    def fit():
        if self.cv == "prefit":
            # `_fit_calibrator` using all X and y
            # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
        else:
            if ensemble:
                for train, test in cv.split(X, y):
                    # fit classifier
                    # call `_fit_calibrator`, producing a `_CalibratedClassiferPipeline` instance
                    # append `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`,
                    # ending up with `n_cv` `_CalibratedClassiferPipeline` instances
            else:
                # Use `cross_val_predict` to get all `preds`
                # call `_fit_calibrator` using `preds` and y
                # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
```
we want to get rid of the last `else` as there is duplication of code there. With the new `_get_prediction_method` we have:

```Python
class CalibratedClassifierCV():

    def fit():
        if self.cv == "prefit":
            # `_fit_calibrator` using all X and y
            # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
        else:
            for train, test in cv.split(X, y):
                # fit classifier
                # call `_fit_calibrator`, which calls `_get_predictions`.
                # In case of `ensemble=False`, we use use `partial(cross_val_predict())`
                # `_fit_calibrator` also fits calibrators and produces `_CalibratedClassiferPipeline` instance
                #  Only one _CalibratedClassiferPipeline` is required when `ensemble=False`
                # (i.e. shouldn't be in the for loop)
                # when `ensemble=True` `n_cv` `_CalibratedClassiferPipeline` instances are needed 
                # (i.e., should be in the for loop) 
```

Reply:
Sorry if I wasn't clear! I don't think we can move away from the first version (with the last `else`). The duplication of code that I wanted to avoid was:

1. getting the right method

```py

                if hasattr(base_estimator, "decision_function"):
                    base_estimator_method = "decision_function"
                elif hasattr(base_estimator, "predict_proba"):
                    base_estimator_method = "predict_proba"
```

2. reshaping the predictions

```py
                if base_estimator_method == "decision_function":
                    if preds.ndim == 1:
                        preds = preds[:, np.newaxis]
                else:
                    if len(self.label_encoder_.classes_) == 2:
                        preds = preds[:, 1]
```

Right now, these 2 things happen both in `CalibratedClassifierCV.fit` (for ensemble=False) and in `_get_predictions` (for ensemble=True).

So to expand on my previous suggestion:


```py
def get_prediction_method(clf, return_string=False):
	# return clf.decision_function or clf.predict_proba or their corresponding string
	# we need strings when using cross_val_predict

def _get_predictions(X, pred_func):
	# pred_func is either a method of the clf, or a partial cross_val_predict
	preds = pred_func(X)
	# reshape preds here
	return preds


class CalibratedClassifierCV():

    def fit():
        if self.cv == "prefit":
            # `_fit_calibrator` using all X and y
            # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
        else:
            if ensemble:
                for train, test in cv.split(X, y):
                    # fit classifier on train
					preds = _get_predictions(test, get_prediction_method(clf, return_string=False)
                    # call `_fit_calibrator`, producing a `_CalibratedClassiferPipeline` instance
                    # append `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`,
                    # ending up with `n_cv` `_CalibratedClassiferPipeline` instances
            else:
				preds = _get_predictions(X, partial(cross_val_predict(..., method=get_prediction_method(clf, return_string=True)))
                # call `_fit_calibrator` using `preds` and y
                # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`

```

I think we might need to modify `_fit_calibrator` to only accept `preds` and not call `_get_predictions`

Reply:
FYI:
> Also is df = df[:, 1:] just equivalent to df = df[:, 1]?

it seems that it is not the same, you get 2d output with `df[:, 1:]`